File: Leetcode Medium/1448CountGoodNodesInBinaryTree.py
#https://leetcode.com/problems/count-good-nodes-in-binary-tree/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        counter = 0
        def helper(node, maxNum):
            nonlocal counter
            if node == None: return
            if node.val >= maxNum:
                maxNum = node.val
                counter += 1
            helper(node.left, maxNum)
            helper(node.right, maxNum)
        helper(root, float('-inf'))
        return counter

File: C++/Leetcode Medium/1448CountGoodNodesInBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>
#include <utility>
#include <algorithm>
using namespace std;
class Solution {
public:
    int goodNodes(TreeNode* root) {
        int res = 0;
        vector<pair<int, TreeNode*>> stack = {{-99999, root}};
        while (!stack.empty()) {
            pair<int, TreeNode*> curr = stack.back();
            stack.pop_back();
            if (curr.second->val >= curr.first) ++res;
            int val = max(curr.first, curr.second->val);
            if (curr.second->left) stack.push_back({val, curr.second->left});
            if (curr.second->right) stack.push_back({val, curr.second->right});
        }
        return res;
    }
};